modules/javafx.graphics/src/main/native-glass/mac/GlassViewDelegate.m

```diff
@@ -845,6 +851,72 @@ - (void)sendJavaGestureEndEvent:(NSEvent *)theEvent
     GLASS_CHECK_EXCEPTION(env);
 }
 
+/*
+ * This method is a replacement for the deprecated beginGestureWithEvent
+ * method, which is no longer delivered to a View by macOS. This
+ * is called for each gesture event to track the beginning of a
+ * gesture using the phase of the event. We call sendJavaGestureBeginEvent
+ * if there are no other gestures active.
+ */
+- (void)maybeBeginGestureWithEvent:(NSEvent *)theEvent withMask:(GestureMaskType)theMask
+{
+    NSEventPhase phase = [theEvent phase];
+    if (phase == NSEventPhaseBegan) {
+        if (gesturesBeganMask == 0) {
+            [self sendJavaGestureBeginEvent:theEvent];
+        }
+        gesturesBeganMask |= theMask;
```

Comment:
Rotate and Magnify gestures can be started and performed together.
If Rotate is already in progress then `gesturesBeganMask` would be set to `4`.
and If we start a Magnify gesture while the Rotate is not ended then `sendJavaGestureBeginEvent` will not be executed for Magnify gesture.
However the application currently receives the synthetic `ZOOM_STARTED` event, which is generated as a result of call to `sendJavaGestureEvent` and not an event generated from here,

A change in inner `if` statement as below can fix the issue,

```
if ((gesturesBeganMask & theMask) == 0) {
    gesturesBeganMask |= theMask;
    [self sendJavaGestureBeginEvent:theEvent]
}
```

Reply:
This is a good thought, and one I had initially considered, but didn't do for two related reasons:
1. The macOS NSEvent code that sends `beginGestureWithEvent` and `endGestureWithEvent` when the JDK is compiled with an older SDK doesn't send a begin call when starting a rotate gesture if a zoom gesture is already active, or vice versa. One goal of this change is to match the previous behavior to minimize risk.
2. More importantly, the existing logic is not expecting nested `sendJavaGestureBeginEvent` / `sendJavaGestureEndEvent` calls, and will not track the state correctly if they are nested. I did a quick test implementing your suggestion and confirmed that it misbehaves.

Worth noting, the logic that synthesizes the begin and end gesture notifications that are sent by the Java gesture code already takes care of the case of switching from zoom to rotate and sends the correct sequence to the event listener.